### PR TITLE
140/best n sampling

### LINKED ISSUE #140: Add Best of N sampling
What's the right place to add best of n sampling and compare its impact to some existing methods?

Some references:
* Discussed in [reward model scaling laws paper](https://arxiv.org/abs/2210.10760),
* OpenAI [blog post](https://openai.com/blog/measuring-goodharts-law),
* Used in [WebGPT](https://arxiv.org/abs/2112.09332)

### PR DESCRIPTION
**Pull request by metric-space**
Attempts to deliver work for https://github.com/lvwerra/trl/issues/140

Fixes #140

## Gist

Showcase usage of best-of-n sampling

### CODE REVIEW

File: examples/best_of_n.ipynb [hunk 1]
```diff
@@ -0,0 +1,680 @@
+{
```
Comment (younesbelkada) [inline examples/best_of_n.ipynb:1]: <p>Maybe add a few sentences to explain the motivation of the notebook and what does it showcases</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/lvwerra/trl/pull/326/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this! 
Can you also add few notes in the notebook and the `README.md` file of examples explaining the demonstration?
From my understanding this is a new rollout strategy that a user can use in replacement of the existing one. So this should be highlighted for better understanding and broader adoption

**Comment by metric-space** [comment]
@younesbelkada  thank you for the review. I took the liberty of shifting the notebook to a directory within `examples` and have included a README.md within said directory. Whenver you can, please let me know what you think of it

**Review by younesbelkada — APPROVED** [review]
Awesome work! Thanks a lot for your contribution

**Comment by natolambert** [comment]
Sorry was a little swamped with travel and stuff last week, but I'm stoked about this @metric-space !

**Comment by metric-space** [comment]
Hey folks, I just realized that the description was edited to say this fixes #140 (implies finality to me) but I think the notebook is a precursor to actual framework code for `best-of-n sampling` as put here 
https://github.com/lvwerra/trl/issues/140#issuecomment-1425431127
> One idea would be to wrap that logic in a helper class like BestOfNSampler, but starting with a notebook sounds like a good way to go!

Happy to put in the work to get said additions in. Just waiting on more PR reviews to see whether I'm in the clear

**Comment by metric-space** [comment]
Hey folks @natolambert and @younesbelkada, hate to poke/ping people but also don't want this development to remain dormant and go cold. Any word on this?

**Comment by lvwerra** [comment]
Looks good to me. Let's merge the notebook and think about `BestOfNSampler` in a separate PR?

**Review by lvwerra — APPROVED** [review]

**Comment by metric-space** [comment]
Thank you folks for this and I appreciate the response. I'll start up a draft PR and conversations on the thread (if needed) around the `BestOfNSampler` shortly

**Pull request merged**